internal/resolver/config_selector.go

```diff
@@ -0,0 +1,72 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package resolver provides internal resolver-related functionality.
+package resolver
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/serviceconfig"
+	"google.golang.org/grpc/resolver"
+)
+
+// ConfigSelector controls what configuration to use for every RPC.
+type ConfigSelector interface {
+	// Selects the configuration for the RPC.
+	SelectConfig(RPCInfo) *RPCConfig
+}
+
+// RPCInfo contains RPC information needed by a ConfigSelector.
+type RPCInfo struct {
+	Context context.Context // Contains headers, application timeout
+	Method  string          // i.e. "/Service/Method"
+}
+
+// RPCConfig describes the configuration to use for each RPC.
+type RPCConfig struct {
+	// The context to use for the remainder of the RPC; can pass info to LB
+	// policy or affect timeout or metadata.
+	Context      context.Context
```

Comment:
Is it possible to avoid storing this `context` here? 
What is expected to be returned in this `Context` that cannot be returned through the `MethodConfig`?
Does it make sense to return a `context` from `SelectConfig`?
Or can we have a method on this type which returns a context?

Reply:
We could return a metadata and some `Attributes` instead.  We'd use the metadata to overwrite the user's metadata stored in their context and pass the attributes to `balancer.Pick`.  It would work, but this is more convenient and more gRPC-Go-like (even if it's, sadly, not very Go-like).

> Does it make sense to return a context from SelectConfig?

This struct is just a bundle of things returned from `SelectConfig` to avoid having many return values.  We'd be trading:

```go
return RPCConfig{Context: context, ...}
```

for

```
return context, RPCConfig{...}
```

I don't think the latter is better just because it avoids putting a context into a struct.

Reply:
Today we support only a limited number of route actions. My concern is that we could be stuffing more and more things into the context without there being a need to do so. We could instead explicitly add fields in this struct for everything that we want the config selector to return.

Reply:
> We could instead explicitly add fields in this struct for everything that we want the config selector to return.

I don't want to codify the *xDS* config selector -> balancer interface within gRPC, as different CS/B combinations may want to pass different data.

Here's another thing we can do if there are strong concerns about `context`:

```go
type RPCInfo struct {
  Metadata func() metadata.MD  // Callback to get metadata, if needed
  // remove context
...
}
```

And

```go
type RPCConfig struct {
  Attributes attributes.Attributes // that's a lot of attributes
  // remove context
...
}
```

This `Attributes` can be added to `balancer.PickInfo`.  Or, potentially, the context already passed to `Pick` as we expect the config selector to do today.  This is a little weird, because a balancer implementer won't really understand this field, since it's only set by config selectors, which external users don't currently have access to.

Reply:
I think here also, we could just add a comment explaining the rationale behind putting the context into the struct. I guess performance (given that this is on the hot path) and ease of use (since this is an internal API) are valid concerns to not go by the book and pass the context in the struct here. Thanks.